### ISSUE TITLE
Traceback when a submission is tokened

### ISSUE THREAD

**Issue by veluca93** [cms-dev/cms#581]
During a contest, when a submission was tokened by a contestant, we obtained the following traceback:

```
2016-03-11 09:15:58,399 - ERROR [Contest,0 642 web::log_exception] Uncaught exception POST /tasks/coupon/submissions/7/token (10.1.2.2)
HTTPServerRequest(protocol='http', host='ioi-server', method='POST', uri='/tasks/coupon/submissions/7/token', version='HTTP/1.1', remote_ip='10.1.2.2', headers={'Origin': 'https://ioi-server', 'Content-Length': 
'0', 'Accept-Language': 'it-IT,it;q=0.8,en-US;q=0.6,en;q=0.4', 'Accept-Encoding': 'gzip, deflate', 'X-Forwarded-For': '10.1.2.2', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q
=0.8', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/44.0.2403.89 Chrome/44.0.2403.89 Safari/537.36', 'Host': 'ioi-server', 'Referer': 'https://ioi-server/
tasks/coupon/submissions?ZMQ4Wt3-fTbZNod__jQ8bljy3zB4BgGH5NLkSo244to.', 'Https': '1', 'Cache-Control': 'max-age=0', 'X-Real-Ip': '10.1.2.2', 'Cookie': 'login="2|1:0|10:1457684156|5:login|68:KFZtYXp6dWNhdG8KcDAKV
mZidW1wbQpwMQpGMTQ1NzY4NDE1Ni4yODg3NDkKdHAyCi4=|e4788371e8204d2158169b7019630a806fe6a22bacc69636f3b1244b5b4de0db"; unread_count="2|1:0|10:1457684156|12:unread_count|4:MA==|9a12f326f61d2956f5bbe8d0ddf020c8f5e4b62
f7082bfe59f03692da8300d0f"', 'Content-Type': 'application/x-www-form-urlencoded'})
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/tornado/web.py", line 1443, in _execute
    result = method(*self.path_args, **self.path_kwargs)
  File "/usr/lib/python2.7/site-packages/tornado/web.py", line 2800, in wrapper
    return method(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/server/util.py", line 196, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/server/contest/handlers/tasksubmission.py", line 639, in post
    submission_id=submission.id)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/io/rpc.py", line 659, in remote_method
    result = self.execute_rpc(method=method, data=data)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/io/rpc.py", line 602, in execute_rpc
    result.set_exception(RPCError("Write failed."))
  File "/usr/lib/python2.7/site-packages/gevent/event.py", line 344, in set_exception
    self._exc_info = (type(exception), exception, dump_traceback(None))
  File "/usr/lib/python2.7/site-packages/gevent/_tblib.py", line 243, in g
    return f(a)
  File "/usr/lib/python2.7/site-packages/gevent/_tblib.py", line 298, in dump_traceback
    return dumps(tb)
TypeError: 'NoneType' object is not callable
2016-03-11 09:15:58,400 - ERROR [Contest,0 642 base::write_error] Uncaught exception (TypeError("'NoneType' object is not callable",)) while processing a request: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/tornado/web.py", line 1443, in _execute
    result = method(*self.path_args, **self.path_kwargs)
  File "/usr/lib/python2.7/site-packages/tornado/web.py", line 2800, in wrapper
    return method(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/server/util.py", line 196, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/server/contest/handlers/tasksubmission.py", line 639, in post
    submission_id=submission.id)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/io/rpc.py", line 659, in remote_method
    result = self.execute_rpc(method=method, data=data)
  File "/usr/lib/python2.7/site-packages/cms-1.3.dev0-py2.7.egg/cms/io/rpc.py", line 602, in execute_rpc
    result.set_exception(RPCError("Write failed."))
  File "/usr/lib/python2.7/site-packages/gevent/event.py", line 344, in set_exception
    self._exc_info = (type(exception), exception, dump_traceback(None))
  File "/usr/lib/python2.7/site-packages/gevent/_tblib.py", line 243, in g
    return f(a)
  File "/usr/lib/python2.7/site-packages/gevent/_tblib.py", line 298, in dump_traceback
    return dumps(tb)
TypeError: 'NoneType' object is not callable
```

**Comment by stefano-maggiolo** [comment]
Did it happen only once, there were other tokens played in the contest? Is this urgent?

**Comment by wil93** [comment]
It's not urgent. It happened sporadically.

**Comment by stefano-maggiolo** [comment]
Thanks, will take a look.

**Comment by lerks** [comment]
The interesting line in the log you report is `result.set_exception(RPCError("Write failed."))`, meaning the RPC socket between CWS and ES was in an unusable state (closed?). Given the date of your report you may have been using gevent 1.1 (which had been released 6 day before; are you using Arch Linux?). If that is the case my best guess is that you may have been hit by the bug we fixed in dc19ea2230f038d1544afe2ff0fea58a9fb943ed, which caused such sockets to be frequently closed. It's somewhat weird that no one else reported such issue, but on the other hand it's fairly rare to trigger (the closing has to happen between the moment the RPC is initiated and the moment the data is actually sent) and it has small impact. In light of that I am closing this issue; please reopen it if the bug shows up again using the current master.
